Add Colorbar Example to Docs and Clarify Ordering of plot_axes (#383)
sunpy/ndcube: docs/visualization.rst

  >>> plt.cla()
  >>> plt.close()

Note that sometimes axis tickmarks are missing.

> This is a caused by a bug in `~astropy.visualization.wcsaes.WCSAxes` whereby the ticks and labels are omitted if the plot extends beyond the valid range of the WCS projection.

Reviewer: This isn't a bug, as I discovered a while ago. It's in general not possible to display the ticks/labels if the spine of an axis is outside the range of the WCS projection.

Reviewer 2: I think there are a few cases where wcsaxes could do better, but yea there are many where it's not a bug.

Reviewer: Do you have any cases in mind? I'd be interested in taking a look and trying to improve things if there are things to be improved.

Author: ```suggestion
This is a caused by a behavior in `~astropy.visualization.wcsaes.WCSAxes` whereby the ticks and labels are omitted if the plot extends beyond the valid range of the WCS projection.
```

Reviewer 2: > Do you have any cases in mind? I'd be interested in taking a look and trying to improve things if there are things to be improved.

Well the 1D case (without thinking about it too much) seems particularly silly, how do you reach the edge of a projection in 1D??

Also I feel like for some of the whole sky projections where the coordinates are *so close* to the edge of the projection WCSAxes could be smarter, but I think this was the case you looked at before?

  >>> plt.cla()
  >>> plt.close()

Adding Colorbars
----------------
Working with the output of `ndcube.NDCube.plot` is the context of matplotlib figures and axes can be a great way of creating more complex plots.
Here we will show two examples of home to add a colorbar.

The first is simple and depends on `matplotlib.pyplot`.

> .. code-block:: python
> 
>   >>> my_cube.plot()  # doctest: +SKIP
>   >>> plt.colorbar()  # doctest: +SKIP

Author: ```suggestion
.. plot::
  :context:
  :include-source:

  >>> ax = my_cube[0].plot()
  >>> cbar = plt.colorbar()
```

The second example shows how to more intricately play with `~ndcube.NDCube` visualizations and matplotlib figures and axes.
This includes adding the output of `ndcube.NDCube.plot` to an existing axes object.

.. plot::
  :context:
  :include-source:

  >>> fig = plt.figure()  # Create a figure
  >>> # Create WCSAxes object and then add the NDCube plot by setting the axes kwarg.
  >>> ax = fig.add_axes([0.1, 0.1, 0.6, 0.6], projection=my_cube[0].wcs)
  >>> ax = my_cube[0].plot(axes=ax)

>   >>> # Create the colorbar axes object and scale it by the image.

Reviewer 2: This works but the fact you have to go to these lengths is a bug. We should be setting current axes and current image like map does so it's easier to do.

Author: To be honest, because I don't use a lot of AIA images in my own work, I'm not sure how this is usually done with `Map`.  It was the first way I found that worked.  If you think there's a simpler way based on the `Map` workflow we can put that here instead as long as it works.

Reviewer 2: Our plot method (for 2D) should include this: https://github.com/sunpy/sunpy/blob/master/sunpy/map/mapbase.py#L2181-L2184

Author: It does.  For 2-D static images at least: https://github.com/sunpy/ndcube/blob/master/ndcube/mixins/plotting.py#L178

Reviewer 2: oh well then a simple call to `plt.colorbar()` should be all you need?

Author: You mean as shown above?

>   >>> fig = plt.figure()  # Create a figure
>   >>> # Create WCSAxes object and then add the NDCube plot by setting the axes kwarg.
>   >>> ax = fig.add_axes([0.1, 0.1, 0.6, 0.6], projection=my_cube[0].wcs)
>   >>> ax = my_cube[0].plot(axes=ax)
>   >>> # Create the colorbar axes object and scale it by the image.
>   >>> cax = fig.add_axes([0.85, 0.1, 0.05, 0.6])
>   >>> im = ax.get_images()[0]  # Retrieve the plot AxesImage by which to scale colorbar.
>   >>> cbar = fig.colorbar(im, cax=cax, label="Intensity")

Reviewer 2: I think this should work:

```suggestion
  >>> fig = plt.figure()  # Create a figure
  >>> # Create WCSAxes object and then add the NDCube plot by setting the axes kwarg.
  >>> ax = plt.subplot(projection=my_cube[0])
  >>> ax = my_cube[0].plot(axes=ax)
  >>> cbar = plt.colorbar(label="Intensity")
```

Reviewer 2: even the `(axes=ax)` is optional.

Reviewer: I like leaving in `axes=ax`, as it's explicit and shows the user how to choose which axes to plot their cube on.

Reviewer: Does `fig.subplot` work instead? Again, then it's more explicit.

Author: I've tried to make the above example as short and easy as possible and the second to me much more explicit.

Reviewer 2: I always feel like `fig.add_axes` is too much for 99% of the time. Why on earth do you want to be manually defining the size of your axes like this? If you really care use gridspec or something.

We could show an example using only the mpl OO API though, if `fig.subplot` does indeed work.

Author: I'm going to leave that for someone who knows about this stuff to make in a separate PR.

  >>> plt.cla()
  >>> plt.close()

Adding Colorbars
----------------
Working with the output of `ndcube.NDCube.plot` is the context of matplotlib figures and axes can be a great way of creating more complex plots.
Here we will show two examples of home to add a colorbar.

The first is simple and depends on `matplotlib.pyplot`.

.. plot::
  :context:
  :include-source:

  >>> ax = my_cube[0].plot()

>   >>> cbar = plt.colorbar()

Author: ```suggestion
  >>> cbar = plt.colorbar()
  >>> plt.show()
```

  >>> plt.cla()
  >>> plt.close()

Adding Colorbars
----------------
Working with the output of `ndcube.NDCube.plot` is the context of matplotlib figures and axes can be a great way of creating more complex plots.
Here we will show two examples of home to add a colorbar.

The first is simple and depends on `matplotlib.pyplot`.

.. plot::
  :context:
  :include-source:

  >>> ax = my_cube[0].plot()

>   >>> cbar = plt.colorbar()

Author: ```suggestion
  >>> cbar = plt.colorbar(orientation="horizontal")
```

  >>> plt.show()
  
.. plot::
  :nofigs:
  >>> plt.clf()
  >>> plt.cla()
  >>> plt.close()

The second example shows how to more intricately play with `~ndcube.NDCube` visualizations and matplotlib figures and axes.
This includes adding the output of `ndcube.NDCube.plot` to an existing axes object.

.. plot::
  :context:
  :include-source:

  >>> fig = plt.figure()  # Create a figure
  >>> # Create WCSAxes object and then add the NDCube plot by setting the axes kwarg.
  >>> ax = fig.add_axes([0.1, 0.1, 0.6, 0.6], projection=my_cube[0].wcs)
  >>> ax = my_cube[0].plot(axes=ax)
  >>> # Create the colorbar axes object and scale it by the image.
  >>> cax = fig.add_axes([0.85, 0.1, 0.05, 0.6])
  >>> im = ax.get_images()[0]  # Retrieve the plot AxesImage by which to scale colorbar.

>   >>> cbar = fig.colorbar(im, cax=cax, label="Intensity")

Author: ```suggestion
  >>> cbar = fig.colorbar(im, cax=cax, label="Intensity")
  >>> plt.show()
```
